### DISCUSSION TITLE
session.expunge(obj) but returning a new expunged copy instead of expunging the original object

### DISCUSSION THREAD

**Discussion by ddorian** [sqlalchemy/sqlalchemy#11859] [Usage Questions]
Hi team,

I have an object in the main "transaction" thread. And then I create many, 10+, other threads, that each one takes this object (or an object in a relationship to this) and does work.

```python
def do_async_3(obj):
    new_session = Session()
    obj2 = new_session.merge(obj)
    # do the work
    ...

# main thread here
obj = Obj()
# spawn all async threads, but don't start none of them
pool.spawn(do_async_3, obj)
pool.spawn(do_async_n, obj.relation_1)
...
# we commit the session because the waiting may be long, the transaction will expire and we get an error
session.commit()
# we start waiting for all threads at the same to capture all errors
pool.join(raise_errors=True)
.....
# continue work with the obj, it will be reloaded and a new transaction started automatically
```

My issue is:
1. If I don't commit before pool.join(), the transaction in main thread will expire with error (my pg doesn't like long transactions).
2. Often, I get an error inside the `do_async_3`, where the object has been expire by the commit(), and it can't do `merge(obj)` (it's in the form of `NoneType' object has no attribute '_sa_iterator'`, probably a combination of model/dataclass/init=False on some fields along with expiration magic).
3. If I do `obj.expunge()`, I will lose the "obj". I have to do expunge in an outer layer and pass this expunged obj to the "main" thread so it can also merge it.

Is there a way to do a `obj2 = session.return_expunged_copy(obj)` without expunging the original?

Regards,
Dorian

**Comment by zzzeek** [comment]
why dont you turn off expire_on_commit on the main Session ?    seems appropriate here

**Reply by ddorian** [reply]
I've had some issues that when the object is expired, on the main thread, some new values that were set on the other threads wouldn't be available. 

I guess I can fix that with expiring_on_commit only on main thread and doing session.refresh(obj) there too.
